File: amidisk/rdb/blocks.py
"""RDB on-disk block structures: RDSK, PART, FSHD, LSEG, BADB.

Port of amitools' amitools/fs/block/rdb/ block classes (RDBlock,
PartitionBlock, FSHeaderBlock, LoadSegBlock) with the same field names,
which in turn follow the AmigaOS NDK <devices/hardblocks.h> layout.
All values are big-endian longwords; every block carries a checksum:
the sum of the first `size` longwords must be zero.
"""

import struct


def checksum_ok(data, num_longs):
    s = 0
    for i in range(num_longs):
        s = (s + struct.unpack_from(">I", data, i * 4)[0]) & 0xFFFFFFFF
    return s == 0


def fix_checksum(data, num_longs, chk_loc=2):
    """Recompute the checksum longword in a mutable buffer."""
    struct.pack_into(">I", data, chk_loc * 4, 0)
    s = 0
    for i in range(num_longs):
        s = (s + struct.unpack_from(">I", data, i * 4)[0]) & 0xFFFFFFFF
    struct.pack_into(">I", data, chk_loc * 4, (-s) & 0xFFFFFFFF)


class RDBBaseBlock:
    """Common RDB block behavior: magic, size, checksum, longword access."""

    magic = b"????"

    def __init__(self, blkdev, blk_num):
        self.blkdev = blkdev
        self.blk_num = blk_num
        self.data = None
        self.valid = False

    # -- raw access -------------------------------------------------
    def _long(self, idx):
        return struct.unpack_from(">I", self.data, idx * 4)[0]

    def _slong(self, idx):
        return struct.unpack_from(">i", self.data, idx * 4)[0]

    def _put_long(self, idx, val):
        struct.pack_into(">I", self.data, idx * 4, val & 0xFFFFFFFF)

    def _bytes(self, off, size):
        return bytes(self.data[off : off + size])

    def _bstr(self, off, max_chars):
        n = self.data[off]
        if n > max_chars:
            n = max_chars
        return bytes(self.data[off + 1 : off + 1 + n])

    def _put_bstr(self, off, max_chars, val):
        val = val[:max_chars]
        self.data[off] = len(val)
        self.data[off + 1 : off + 1 + len(val)] = val
        # zero pad the rest of the field
        for i in range(off + 1 + len(val), off + 1 + max_chars):
            self.data[i] = 0

    @classmethod
    def new(cls, blkdev, blk_num, size_longs=64):
        """Create a fresh, zeroed block of this type in memory."""
        blk = cls(blkdev, blk_num)
        blk.data = bytearray(blkdev.block_bytes)
        blk.data[0:4] = cls.magic
        blk.size_longs = size_longs
        struct.pack_into(">I", blk.data, 4, size_longs)
        blk.valid = True
        return blk

    # -- I/O ---------------------------------------------------------
    def read(self):
        raw = self.blkdev.read(self.blk_num)
        self.data = bytearray(raw)
        self.valid = False
        if self.data[0:4] != self.magic:
            return False
        self.size_longs = self._long(1)
        if self.size_longs < 3 or self.size_longs * 4 > len(self.data):
            return False
        if not checksum_ok(self.data, self.size_longs):
            return False
        self.valid = True
        self._unpack()
        return True

    def write(self):
        self._pack()
        fix_checksum(self.data, self.size_longs)
        self.blkdev.write(self.blk_num, bytes(self.data))

    def _unpack(self):
        pass

    def _pack(self):
        pass


class RDBlock(RDBBaseBlock):
    """RDSK -- Rigid Disk Block, anchor of the partition table."""

    magic = b"RDSK"

    # rb_Flags
    FLAG_LAST = 0x01
    FLAG_LAST_LUN = 0x02
    FLAG_LAST_TID = 0x04
    FLAG_NO_RESELECT = 0x08
    FLAG_DISK_ID = 0x10
    FLAG_CTRL_ID = 0x20
    FLAG_SYNCH = 0x40

    def _unpack(self):
        self.host_id = self._long(3)
        self.block_bytes = self._long(4)
        self.flags = self._long(5)
        self.badblock_list = self._long(6)
        self.partition_list = self._long(7)
        self.fs_header_list = self._long(8)
        self.drive_init = self._long(9)
        # physical drive characteristics
        self.cylinders = self._long(16)
        self.sectors = self._long(17)
        self.heads = self._long(18)
        self.interleave = self._long(19)
        self.park = self._long(20)
        self.write_precomp = self._long(24)
        self.reduced_write = self._long(25)
        self.step_rate = self._long(26)
        # logical drive characteristics
        self.rdb_blocks_lo = self._long(32)
        self.rdb_blocks_hi = self._long(33)
        self.lo_cylinder = self._long(34)
        self.hi_cylinder = self._long(35)
        self.cyl_blocks = self._long(36)
        self.auto_park_seconds = self._long(37)
        self.high_rdsk_block = self._long(38)
        # drive identification
        self.disk_vendor = self._bytes(160, 8).rstrip(b"\x00 ").decode("latin-1")
        self.disk_product = self._bytes(168, 16).rstrip(b"\x00 ").decode("latin-1")
        self.disk_revision = self._bytes(184, 4).rstrip(b"\x00 ").decode("latin-1")
        self.ctrl_vendor = self._bytes(188, 8).rstrip(b"\x00 ").decode("latin-1")
        self.ctrl_product = self._bytes(196, 16).rstrip(b"\x00 ").decode("latin-1")
        self.ctrl_revision = self._bytes(212, 4).rstrip(b"\x00 ").decode("latin-1")

    def _put_padded(self, off, size, text):
        raw = text.encode("latin-1")[:size].ljust(size, b" ")
        self.data[off : off + size] = raw

    def _pack(self):
        self._put_long(3, self.host_id)
        self._put_long(4, self.block_bytes)
        self._put_long(5, self.flags)
        self._put_long(6, self.badblock_list)
        self._put_long(7, self.partition_list)
        self._put_long(8, self.fs_header_list)
        self._put_long(9, self.drive_init)
        for i in range(10, 16):
            self._put_long(i, 0xFFFFFFFF)
        self._put_long(16, self.cylinders)
        self._put_long(17, self.sectors)
        self._put_long(18, self.heads)
        self._put_long(19, self.interleave)
        self._put_long(20, self.park)
        self._put_long(24, self.write_precomp)
        self._put_long(25, self.reduced_write)
        self._put_long(26, self.step_rate)
        self._put_long(32, self.rdb_blocks_lo)
        self._put_long(33, self.rdb_blocks_hi)
        self._put_long(34, self.lo_cylinder)
        self._put_long(35, self.hi_cylinder)
        self._put_long(36, self.cyl_blocks)
        self._put_long(37, self.auto_park_seconds)
        self._put_long(38, self.high_rdsk_block)
        self._put_padded(160, 8, self.disk_vendor)
        self._put_padded(168, 16, self.disk_product)
        self._put_padded(184, 4, self.disk_revision)
        self._put_padded(188, 8, self.ctrl_vendor)
        self._put_padded(196, 16, self.ctrl_product)
        self._put_padded(212, 4, self.ctrl_revision)


class DosEnvec:
    """de_* environment vector embedded in PART blocks (hardblocks.h)."""

    FIELDS = (
        "table_size",   # highest valid index below
        "size_block",   # block size in longwords
        "sec_org",
        "surfaces",
        "sec_per_blk",  # sectors per filesystem block
        "blk_per_trk",
        "reserved",     # blocks reserved at partition start (bootblocks)
        "pre_alloc",
        "interleave",
        "low_cyl",
        "high_cyl",
        "num_buffer",
        "buf_mem_type",
        "max_transfer",
        "mask",
        "boot_pri",
        "dos_type",
        "baud",
        "control",
        "boot_blocks",
    )

    def __init__(self):
        for f in self.FIELDS:
            setattr(self, f, 0)

    @classmethod
    def parse(cls, blk, base_long):
        env = cls()
        table_size = blk._long(base_long)
        env.table_size = table_size
        for i, f in enumerate(cls.FIELDS[1:], start=1):
            if i <= table_size:
                setattr(env, f, blk._long(base_long + i))
        env.boot_pri = blk._slong(base_long + 15) if table_size >= 15 else 0
        return env

    def pack(self, blk, base_long):
        blk._put_long(base_long, self.table_size)
        for i, f in enumerate(self.FIELDS[1:], start=1):
            if i <= self.table_size:
                blk._put_long(base_long + i, getattr(self, f))

    # -- derived geometry --------------------------------------------
    def cyl_secs(self):
        return self.surfaces * self.blk_per_trk

    def num_cyls(self):
        return self.high_cyl - self.low_cyl + 1

    def start_sec(self):
        return self.low_cyl * self.cyl_secs()

    def num_secs(self):
        return self.num_cyls() * self.cyl_secs()

    def dos_type_str(self):
        return dos_type_to_str(self.dos_type)


def dos_type_to_str(dt):
    tag = struct.pack(">I", dt)
    txt = ""
    for b in tag[:3]:
        txt += chr(b) if 32 <= b < 127 else "?"
    return "%s\\%X" % (txt, tag[3])


class PartitionBlock(RDBBaseBlock):
    """PART -- one partition definition."""

    magic = b"PART"

    FLAG_BOOTABLE = 0x01
    FLAG_NO_MOUNT = 0x02

    def _unpack(self):
        self.host_id = self._long(3)
        self.next = self._long(4)
        self.flags = self._long(5)
        self.dev_flags = self._long(8)
        self.drv_name = self._bstr(36, 31).decode("latin-1")
        self.dos_env = DosEnvec.parse(self, 32)

    def _pack(self):
        self._put_long(3, self.host_id)
        self._put_long(4, self.next)
        self._put_long(5, self.flags)
        self._put_long(8, self.dev_flags)
        self._put_bstr(36, 31, self.drv_name.encode("latin-1"))
        self.dos_env.pack(self, 32)

    def bootable(self):
        return bool(self.flags & self.FLAG_BOOTABLE)

    def no_mount(self):
        return bool(self.flags & self.FLAG_NO_MOUNT)


class FSHeaderBlock(RDBBaseBlock):
    """FSHD -- embedded filesystem driver header."""

    magic = b"FSHD"

    def _unpack(self):
        self.host_id = self._long(3)
        self.next = self._long(4)
        self.flags = self._long(5)
        self.dos_type = self._long(8)
        self.version = self._long(9)
        self.patch_flags = self._long(10)
        # DeviceNode fields
        self.dn_type = self._long(11)
        self.dn_task = self._long(12)
        self.dn_lock = self._long(13)
        self.dn_handler = self._long(14)
        self.dn_stack_size = self._long(15)
        self.dn_priority = self._long(16)
        self.dn_startup = self._long(17)
        self.dn_seg_list_blk = self._long(18)
        self.dn_global_vec = self._long(19)

    def version_tuple(self):
        return (self.version >> 16, self.version & 0xFFFF)

    def dos_type_str(self):
        return dos_type_to_str(self.dos_type)


class LoadSegBlock(RDBBaseBlock):
    """LSEG -- chained chunks of a filesystem driver binary (hunk data)."""

    magic = b"LSEG"

    def _unpack(self):
        self.host_id = self._long(3)
        self.next = self._long(4)
        self.load_data = self._bytes(20, (self.size_longs - 5) * 4)


class BadBlocksBlock(RDBBaseBlock):
    """BADB -- bad block remapping (rare; preserved but not interpreted)."""

    magic = b"BADB"

    def _unpack(self):
        self.host_id = self._long(3)
        self.next = self._long(4)

File: amidisk_cpp/src/rdb/blocks.cpp
#include "blocks.h"
#include <cstring>
#include <map>
#include <algorithm>

namespace amidisk {

bool checksum_ok(std::span<const uint8_t> data, uint32_t num_longs) {
    if (data.size() < num_longs * 4) return false;
    uint32_t sum = 0;
    for (uint32_t i = 0; i < num_longs; ++i) {
        sum += read_be32(data, i * 4);
    }
    return sum == 0;
}

void fix_checksum(std::span<uint8_t> data, uint32_t num_longs, uint32_t chk_loc) {
    if (data.size() < num_longs * 4) return;
    write_be32(data, chk_loc * 4, 0);
    uint32_t sum = 0;
    for (uint32_t i = 0; i < num_longs; ++i) {
        sum += read_be32(data, i * 4);
    }
    write_be32(data, chk_loc * 4, static_cast<uint32_t>(-static_cast<int32_t>(sum)));
}

std::string dos_type_to_str(uint32_t dt) {
    static const std::map<uint32_t, std::string> known = {
        {0x444f5300, "DOS\\0 (OFS Non-intl)"},
        {0x444f5301, "DOS\\1 (FFS Non-intl)"},
        {0x444f5302, "DOS\\2 (OFS-Intl)"},
        {0x444f5303, "DOS\\3 (FFS-Intl)"},
        {0x444f5304, "DOS\\4 (OFS-DirCache)"},
        {0x444f5305, "DOS\\5 (FFS-DirCache)"},
        {0x444f5306, "DOS\\6 (OFS-LNFS)"},
        {0x444f5307, "DOS\\7 (FFS-LNFS)"},
        {0x53465300, "SFS\\0 (Smart File System v1)"},
        {0x53465302, "SFS\\2 (Smart File System v2)"},
        {0x50465300, "PFS\\0 (PFS/PFS2 Floppy/Basic)"},
        {0x50465301, "PFS\\1 (PFS Hard disk)"},
        {0x50465302, "PFS\\2 (PFS3 Legacy)"},
        {0x50465303, "PFS\\3 (PFS3)"},
        {0x50465333, "PFS3 (PFS3 Modern Standard)"},
        {0x4a584604, "JXF\\4 (JXFS AmigaOS 4/MorphOS)"},
        {0x53574150, "SWAP (Amiga Virtual/Swap)"},
        {0x4d414300, "MAC\\0 (Mac OS)"},
        {0x4c4e5800, "LNX\\0 (Linux native)"},
        {0x53575000, "SWP\\0 (Linux swap)"},
        {0x46415400, "FAT\\0 (FAT12/16)"},
        {0x46415401, "FAT\\1 (FAT32)"},
        {0x4e544653, "NTFS (Windows NTFS)"},
        {0x65784654, "exFT (exFAT)"},
        {0x45585434, "EXT4 (Linux EXT4)"}
    };
    auto it = known.find(dt);
    if (it != known.end()) return it->second;

    std::string txt = "";
    uint8_t tag[4];
    write_be32(tag, 0, dt);
    for (int i = 0; i < 3; ++i) {
        if (tag[i] >= 32 && tag[i] < 127) {
            txt += static_cast<char>(tag[i]);
        } else {
            txt += '?';
        }
    }
    char buf[32];
    std::snprintf(buf, sizeof(buf), "%s\\%X", txt.c_str(), tag[3]);
    return buf;
}

// RDBBaseBlock
RDBBaseBlock::RDBBaseBlock(std::shared_ptr<BlockDevice> blkdev, uint64_t blk_num)
    : blkdev_(blkdev), blk_num_(blk_num), valid_(false), size_longs_(0) {}

void RDBBaseBlock::init_new(uint32_t size_longs) {
    data_.resize(blkdev_->sector_size(), 0);
    const char* magic = get_magic();
    std::memcpy(data_.data(), magic, 4);
    size_longs_ = size_longs;
    put_long(1, size_longs_);
    valid_ = true;
}

bool RDBBaseBlock::read() {
    data_.resize(blkdev_->sector_size());
    try {
        blkdev_->read(blk_num_ * blkdev_->sector_size(), data_);
    } catch (const BlockDeviceError&) {
        return false;
    }
    valid_ = false;
    
    if (std::memcmp(data_.data(), get_magic(), 4) != 0) return false;
    
    size_longs_ = get_long(1);
    if (size_longs_ < 3 || size_longs_ * 4 > data_.size()) return false;
    if (!checksum_ok(data_, size_longs_)) return false;
    
    valid_ = true;
    unpack();
    return true;
}

void RDBBaseBlock::write() {
    pack();
    fix_checksum(data_, size_longs_);
    blkdev_->write(blk_num_ * blkdev_->sector_size(), data_);
}

uint32_t RDBBaseBlock::get_long(uint32_t idx) const {
    return read_be32(data_, idx * 4);
}

int32_t RDBBaseBlock::get_slong(uint32_t idx) const {
    return static_cast<int32_t>(read_be32(data_, idx * 4));
}

void RDBBaseBlock::put_long(uint32_t idx, uint32_t val) {
    write_be32(data_, idx * 4, val);
}

void RDBBaseBlock::put_slong(uint32_t idx, int32_t val) {
    write_be32(data_, idx * 4, static_cast<uint32_t>(val));
}

std::string RDBBaseBlock::get_bytes(uint32_t off, uint32_t size) const {
    return std::string(reinterpret_cast<const char*>(&data_[off]), size);
}

std::string RDBBaseBlock::get_bstr(uint32_t off, uint32_t max_chars) const {
    uint8_t len = std::min<uint8_t>(data_[off], static_cast<uint8_t>(max_chars));
    return std::string(reinterpret_cast<const char*>(&data_[off + 1]), len);
}

void RDBBaseBlock::put_bstr(uint32_t off, uint32_t max_chars, const std::string& val) {
    uint32_t len = std::min<uint32_t>(val.length(), max_chars);
    data_[off] = static_cast<uint8_t>(len);
    std::memcpy(&data_[off + 1], val.data(), len);
    std::memset(&data_[off + 1 + len], 0, max_chars - len);
}

void RDBBaseBlock::put_padded(uint32_t off, uint32_t size, const std::string& text) {
    uint32_t len = std::min<uint32_t>(text.length(), size);
    std::memcpy(&data_[off], text.data(), len);
    std::memset(&data_[off + len], ' ', size - len);
}

// RDBlock
std::unique_ptr<RDBlock> RDBlock::create(std::shared_ptr<BlockDevice> blkdev, uint64_t blk_num) {
    auto b = std::make_unique<RDBlock>(blkdev, blk_num);
    b->init_new(64);
    return b;
}

void RDBlock::unpack() {
    host_id = get_long(3);
    block_bytes = get_long(4);
    flags = get_long(5);
    badblock_list = get_long(6);
    partition_list = get_long(7);
    fs_header_list = get_long(8);
    drive_init = get_long(9);
    
    cylinders = get_long(16);
    sectors = get_long(17);
    heads = get_long(18);
    interleave = get_long(19);
    park = get_long(20);
    write_precomp = get_long(24);
    reduced_write = get_long(25);
    step_rate = get_long(26);

    rdb_blocks_lo = get_long(32);
    rdb_blocks_hi = get_long(33);
    lo_cylinder = get_long(34);
    hi_cylinder = get_long(35);
    cyl_blocks = get_long(36);
    auto_park_seconds = get_long(37);
    high_rdsk_block = get_long(38);

    auto trim_spaces = [](std::string s) {
        s.erase(s.find_last_not_of(" \0") + 1);
        return s;
    };
    disk_vendor = trim_spaces(get_bytes(160, 8));
    disk_product = trim_spaces(get_bytes(168, 16));
    disk_revision = trim_spaces(get_bytes(184, 4));
    ctrl_vendor = trim_spaces(get_bytes(188, 8));
    ctrl_product = trim_spaces(get_bytes(196, 16));
    ctrl_revision = trim_spaces(get_bytes(212, 4));
}

void RDBlock::pack() {
    put_long(3, host_id);
    put_long(4, block_bytes);
    put_long(5, flags);
    put_long(6, badblock_list);
    put_long(7, partition_list);
    put_long(8, fs_header_list);
    put_long(9, drive_init);
    for (int i = 10; i < 16; ++i) put_long(i, 0xFFFFFFFF);
    put_long(16, cylinders);
    put_long(17, sectors);
    put_long(18, heads);
    put_long(19, interleave);
    put_long(20, park);
    put_long(24, write_precomp);
    put_long(25, reduced_write);
    put_long(26, step_rate);
    put_long(32, rdb_blocks_lo);
    put_long(33, rdb_blocks_hi);
    put_long(34, lo_cylinder);
    put_long(35, hi_cylinder);
    put_long(36, cyl_blocks);
    put_long(37, auto_park_seconds);
    put_long(38, high_rdsk_block);
    put_padded(160, 8, disk_vendor);
    put_padded(168, 16, disk_product);
    put_padded(184, 4, disk_revision);
    put_padded(188, 8, ctrl_vendor);
    put_padded(196, 16, ctrl_product);
    put_padded(212, 4, ctrl_revision);
}

// DosEnvec
void DosEnvec::parse(const RDBBaseBlock* blk, uint32_t base_long) {
    table_size = blk->get_long(base_long);
    if (table_size >= 1) size_block = blk->get_long(base_long + 1);
    if (table_size >= 2) sec_org = blk->get_long(base_long + 2);
    if (table_size >= 3) surfaces = blk->get_long(base_long + 3);
    if (table_size >= 4) sec_per_blk = blk->get_long(base_long + 4);
    if (table_size >= 5) blk_per_trk = blk->get_long(base_long + 5);
    if (table_size >= 6) reserved = blk->get_long(base_long + 6);
    if (table_size >= 7) pre_alloc = blk->get_long(base_long + 7);
    if (table_size >= 8) interleave = blk->get_long(base_long + 8);
    if (table_size >= 9) low_cyl = blk->get_long(base_long + 9);
    if (table_size >= 10) high_cyl = blk->get_long(base_long + 10);
    if (table_size >= 11) num_buffer = blk->get_long(base_long + 11);
    if (table_size >= 12) buf_mem_type = blk->get_long(base_long + 12);
    if (table_size >= 13) max_transfer = blk->get_long(base_long + 13);
    if (table_size >= 14) mask = blk->get_long(base_long + 14);
    if (table_size >= 15) boot_pri = blk->get_slong(base_long + 15);
    if (table_size >= 16) dos_type = blk->get_long(base_long + 16);
    if (table_size >= 17) baud = blk->get_long(base_long + 17);
    if (table_size >= 18) control = blk->get_long(base_long + 18);
    if (table_size >= 19) boot_blocks = blk->get_long(base_long + 19);
}

void DosEnvec::pack(RDBBaseBlock* blk, uint32_t base_long) const {
    blk->put_long(base_long, table_size);
    if (table_size >= 1) blk->put_long(base_long + 1, size_block);
    if (table_size >= 2) blk->put_long(base_long + 2, sec_org);
    if (table_size >= 3) blk->put_long(base_long + 3, surfaces);
    if (table_size >= 4) blk->put_long(base_long + 4, sec_per_blk);
    if (table_size >= 5) blk->put_long(base_long + 5, blk_per_trk);
    if (table_size >= 6) blk->put_long(base_long + 6, reserved);
    if (table_size >= 7) blk->put_long(base_long + 7, pre_alloc);
    if (table_size >= 8) blk->put_long(base_long + 8, interleave);
    if (table_size >= 9) blk->put_long(base_long + 9, low_cyl);
    if (table_size >= 10) blk->put_long(base_long + 10, high_cyl);
    if (table_size >= 11) blk->put_long(base_long + 11, num_buffer);
    if (table_size >= 12) blk->put_long(base_long + 12, buf_mem_type);
    if (table_size >= 13) blk->put_long(base_long + 13, max_transfer);
    if (table_size >= 14) blk->put_long(base_long + 14, mask);
    if (table_size >= 15) blk->put_slong(base_long + 15, boot_pri);
    if (table_size >= 16) blk->put_long(base_long + 16, dos_type);
    if (table_size >= 17) blk->put_long(base_long + 17, baud);
    if (table_size >= 18) blk->put_long(base_long + 18, control);
    if (table_size >= 19) blk->put_long(base_long + 19, boot_blocks);
}

// PartitionBlock
std::unique_ptr<PartitionBlock> PartitionBlock::create(std::shared_ptr<BlockDevice> blkdev, uint64_t blk_num) {
    auto b = std::make_unique<PartitionBlock>(blkdev, blk_num);
    b->init_new(64);
    return b;
}

void PartitionBlock::unpack() {
    host_id = get_long(3);
    next = get_long(4);
    flags = get_long(5);
    
    // Read up to 31 chars
    drive_name = get_bstr(36, 31);
    
    // Parse DosEnvec at offset 32 longs
    envec.parse(this, 32);
}

void PartitionBlock::pack() {
    put_long(3, host_id);
    put_long(4, next);
    put_long(5, flags);
    
    // Initialize reserved fields
    for (int i = 6; i < 9; ++i) put_long(i, 0); // reserved
    for (int i = 10; i < 32; ++i) put_long(i, 0); // padding around drive_name
    
    put_bstr(36, 31, drive_name);
    
    envec.pack(this, 32);
}

// FSHeaderBlock
std::unique_ptr<FSHeaderBlock> FSHeaderBlock::create(std::shared_ptr<BlockDevice> blkdev, uint64_t blk_num) {
    auto b = std::make_unique<FSHeaderBlock>(blkdev, blk_num);
    b->init_new(64);
    return b;
}

void FSHeaderBlock::unpack() {
    host_id = get_long(3);
    next = get_long(4);
    flags = get_long(5);
    version = (get_long(8) >> 16) & 0xFFFF;
    revision = get_long(8) & 0xFFFF;
    patch_flags = get_long(9);
    type = get_long(10);
    task = get_long(11);
    lock = get_long(12);
    handler = get_long(13);
    stack_size = get_long(14);
    priority = get_slong(15);
    startup = get_slong(16);
    seg_list_blk = get_slong(17);
    global_vec = get_slong(18);
}

void FSHeaderBlock::pack() {
    put_long(3, host_id);
    put_long(4, next);
    put_long(5, flags);
    
    put_long(6, 0); put_long(7, 0); // reserved
    
    put_long(8, (static_cast<uint32_t>(version) << 16) | revision);
    put_long(9, patch_flags);
    put_long(10, type);
    put_long(11, task);
    put_long(12, lock);
    put_long(13, handler);
    put_long(14, stack_size);
    put_slong(15, priority);
    put_slong(16, startup);
    put_slong(17, seg_list_blk);
    put_slong(18, global_vec);
    
    for (int i = 19; i < 64; ++i) put_long(i, 0);
}

// LoadSegBlock
std::unique_ptr<LoadSegBlock> LoadSegBlock::create(std::shared_ptr<BlockDevice> blkdev, uint64_t blk_num) {
    auto b = std::make_unique<LoadSegBlock>(blkdev, blk_num);
    b->init_new(128); // Standard size is often larger
    return b;
}

void LoadSegBlock::unpack() {
    host_id = get_long(3);
    next = get_long(4);
    
    // load_data starts at longword 5, length is size_longs - 5
    uint32_t len_longs = size_longs_ - 5;
    load_data.resize(len_longs * 4);
    
    // Access data_ directly
    std::memcpy(load_data.data(), &data_[20], len_longs * 4);
}

void LoadSegBlock::pack() {
    put_long(3, host_id);
    put_long(4, next);
    
    uint32_t len_longs = (load_data.size() + 3) / 4;
    size_longs_ = 5 + len_longs;
    data_.resize(size_longs_ * 4, 0);
    put_long(1, size_longs_);
    
    std::memcpy(&data_[20], load_data.data(), load_data.size());
}

// BadBlocksBlock
std::unique_ptr<BadBlocksBlock> BadBlocksBlock::create(std::shared_ptr<BlockDevice> blkdev, uint64_t blk_num) {
    auto b = std::make_unique<BadBlocksBlock>(blkdev, blk_num);
    b->init_new(64);
    return b;
}

void BadBlocksBlock::unpack() {
    host_id = get_long(3);
    next = get_long(4);
}

void BadBlocksBlock::pack() {
    put_long(3, host_id);
    put_long(4, next);
    put_long(5, 0); // reserved
}

} // namespace amidisk
